core/src/main/java/com/linecorp/armeria/common/stream/ByteBufDeframerInput.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.util.ReferenceCountUtil;
+
+final class ByteBufDeframerInput implements HttpDeframerInput {
+
+    private final ByteBufAllocator alloc;
+    private final Queue<ByteBuf> queue;
+
+    ByteBufDeframerInput(ByteBufAllocator alloc) {
+        this.alloc = alloc;
+        queue = new ArrayDeque<>();
+    }
+
+    void add(ByteBuf byteBuf) {
+        queue.add(byteBuf);
+    }
+
+    @Override
+    public int readableBytes() {
+        if (queue.isEmpty()) {
+            return 0;
+        }
+
+        int readableBytes = 0;
+        for (ByteBuf buf : queue) {
+            readableBytes += buf.readableBytes();
+        }
+        return readableBytes;
+    }
+
+    private int readerIndex() {
+        if (queue.isEmpty()) {
+            return 0;
+        }
+
+        int value = 0;
+        for (ByteBuf buf : queue) {
+            value += buf.readerIndex();
+        }
+        return value;
+    }
+
+    private int writerIndex() {
+        if (queue.isEmpty()) {
+            return 0;
+        }
+
+        int value = 0;
+        for (ByteBuf buf : queue) {
+            value += buf.writerIndex();
+        }
+        return value;
+    }
```

Comment:
How these two methods calculate `reader/writerIndex` is not correct in my opinion. Actually, there's no notion of reader/writer index from the caller's perspective because it's an `HttpDeframerInput`, not a `ByteBuf`.